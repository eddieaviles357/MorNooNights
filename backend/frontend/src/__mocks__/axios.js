'use strict';

module.exports = {
  getCurrentUser: () => {
    return Promise.resolve({
      data: {
        createdAt: "2023-09-07T03:49:45.893Z",
        email: "eddie@aviles.com",
        firstName: "eddie",
        isAdmin: false,
        lastName: "aviles",
        username: "eddie",
      }
    })
  },
  getRecents: () => {
    return Promise.resolve({
      data: [
        {
          uuid: 'f163aa79-07a0-4673-8752-902323cd2685',
          title: 'Simone Biles Shows She’s Not Just Easing Her Way Back',
          description: 'In just her second elite meet following a two-year hiatus, Biles, 26, won a record eighth U.S. all-around title and became the oldest gymnast to win the champio...',
          keywords: '',
          snippet: 'That perhaps could partially be chalked up to her attitude: “It’s just gymnastics,” she has told her younger teammates. Keeping with her relaxed demeanor,...',
          url: 'https://www.nytimes.com/2023/08/27/sports/simone-biles-gymnastics-championships.html',
          image_url: 'https://static01.nyt.com/images/2023/08/27/multimedia/27gym-biles-top-cbjv/27gym-biles-top-cbjv-facebookJumbo.jpg',
          language: 'en',
          published_at: '2023-01-28T02:57:13.000000Z',
          visited_at: '2023-01-28T02:57:13.000000Z',
          source: 'nytimes.com',
          locale: 'us'
        },
        {
          uuid: 'bce0958b-d645-417c-8a81-c571719af8a6',
          title: 'Simone Biles wins a record 8th U.S. gymnastics title a full decade after her first',
          description: 'Biles became the oldest woman to win a national title since USA Gymnastics began organizing the event in 1963. Her eight crowns moved her past Alfred Jochim, wh...',
          keywords: '',
          snippet: 'Simone Biles wins a record 8th U.S. gymnastics title a full decade after her first\n' +
            '\n' +
            'Enlarge this image toggle caption Godofredo A. Vásquez/AP Godofredo A. Vás...',
          url: 'https://www.npr.org/2023/08/27/1196274503/simone-biles-wins-a-record-8th-u-s-gymnastics-title-a-full-decade-after-her-firs',
          image_url: 'https://media.npr.org/assets/img/2023/08/27/ap23239779365476_wide-91818d99986b4dc474aa1ff648773af3110113b0-s1400-c100.jpg',
          language: 'en',
          published_at: '2023-01-28T02:51:07.000000Z',
          visited_at: '2023-01-28T02:51:07.000000Z',
          source: 'npr.org',
          locale: 'us'
        },
        {
          uuid: '57f3339d-71e5-4a21-b9ed-6e4a31260f57',
          title: 'Ron DeSantis booed at vigil as hundreds mourn victims of racially-motivated Jacksonville shooting',
          description: 'Hundreds of people gathered Sunday at prayer vigils and in church to mourn the three people killed in a racially-motivated shooting at a Dollar General in Jacks...',
          keywords: 'News, florida, jacksonville, mass shootings, racism, Ron DeSantis, shootings',
          snippet: 'JACKSONVILLE, Fla. — Hundreds of people gathered Sunday at prayer vigils and in church, in frustration and exhaustion, to mourn yet another racist attack in A...',
          url: 'https://nypost.com/2023/08/27/ron-desantis-booed-at-vigil-for-victims-of-racially-motivated-jacksonville-shooting/',
          image_url: 'https://nypost.com/wp-content/uploads/sites/2/2023/08/newspress-collage-l3r85vx4q-1693187699533.jpg?quality=75&strip=all&1693173461&w=1024',
          language: 'en',
          published_at: '2023-01-28T02:44:05.000000Z',
          visited_at: '2023-01-28T02:44:05.000000Z',
          source: 'nypost.com',
          locale: 'us'
        }
      ]
    })
  }
}