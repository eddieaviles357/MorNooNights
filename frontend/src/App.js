import React, { useState, useEffect } from 'react';
import Navigation from "./components/routes-nav/Navigation"
import AppRoutes from "./components/routes-nav/Routes"
import useLocalStorage from "./components/hooks/useLocalStorage";
import UserContext from "./components/auth/UserContext";
import { BrowserRouter } from 'react-router-dom';
import MorNooNightsNewsAPI from "./api/api";
import jwt from "jsonwebtoken";
import "./App.css";

// Used for storing token in localStorage for "remember me" re-login
export const TOKEN_STORAGE_ID = "mornoonightsnews-token";

function App() {
  const [infoLoaded, setInfoLoaded] = useState(false);
  const [currentUser, setCurrentUser] = useState(null);
  const [token, setToken] = useLocalStorage(TOKEN_STORAGE_ID);
  const [news, setNews] = useState(null);
  const [visitedNews, setVisitedNews] = useLocalStorage("recents");
  const [topNews, setTopNews] = useState([]);
  const [errors, setErrors] = useState(null);
  
  useEffect(() => {
    console.debug("App useEffect loadUserInfo", "token=", token);

    async function getCurrentUser() {
      if (token) {
        try {
          let { username } = jwt.decode(token);
          // add token to api class so we can interact with
          MorNooNightsNewsAPI.token = token;
          let currentUser = await MorNooNightsNewsAPI.getCurrentUser(username);
          setCurrentUser(currentUser);
        } catch (err) {
          console.error("App loadUserInfo: problem loading", err);
          setCurrentUser(null);
        }
      };
      setInfoLoaded(true);
    }
    setInfoLoaded(false);
    getCurrentUser();
  }, [token]);

  useEffect(() => {
    async function fetchRecents() {
      if(token && currentUser) {
        try {
          const {username} = currentUser;
          const {recents} = await MorNooNightsNewsAPI.getRecents( username ) || [];
          setVisitedNews( JSON.stringify(recents) );
        } catch (err) {
          setErrors(Array.from(err || err.message));
          console.log(err);
        };
      }
    };
    fetchRecents();
  }, [currentUser]);

  // Handles site logout.
  async function logout() {
    try {
      const { username } = currentUser;
      const data = { recents: JSON.parse(visitedNews) };
      console.log("__LOGGIING_OUT__::", data)
      await MorNooNightsNewsAPI.updateRecents(username, data);
    } catch (errors) {
      alert("failed to update recents", errors);
    }
    setCurrentUser(null);
    setToken(null);
    setVisitedNews(null);
    // update recent news
  }

  /** Handles site signup.
 *
 * Automatically logs them in (set token) upon signup.
 *
 */
  async function signup(signupData) {
    try {
      let token = await MorNooNightsNewsAPI.signup(signupData);
      setToken(token);
      return { success: true };
    } catch (errors) {
      console.error("signup failed", errors);
      return { success: false, errors };
    }
  };

  // Handles site login.

  async function login(loginData) {
    try {
      let token = await MorNooNightsNewsAPI.login(loginData);
      setToken(token);
      return { success: true };
    } catch (errors) {
      console.error("login failed", errors);
      return { success: false, errors };
    }
  };

  // handles visited news update to db
  const updateRecentlyVisited = (news) => {
    // if no data is available set to empty array
    const visited = JSON.parse(visitedNews) || [];
    const updatedNews = {...news, visited_at: new Date().toDateString() }
    // allows only 5 entries
    if(visited.length >= 5) visited.shift(); // remove 1st entry
    visited.push(updatedNews)
    // console.log("__HANDLE_CLICK__", updatedNews)
    // stringify visited news and update values
    setVisitedNews( _ => JSON.stringify(visited) )
  }

  if (!infoLoaded) return <div className="text-center">Loading...</div>;

  return (
    <div className="App">
    <BrowserRouter>
      <UserContext.Provider value={{ 
        currentUser, 
        setCurrentUser, 
        visitedNews, 
        setVisitedNews,
        updateRecentlyVisited,
        news,
        setNews,
        topNews,
        setTopNews,
        errors,
        setErrors,
        }}>
        <Navigation logout={logout}/>
        <AppRoutes login={login} signup={signup}/>
      </UserContext.Provider>
    </BrowserRouter>
    </div>
  );
}

export default App;