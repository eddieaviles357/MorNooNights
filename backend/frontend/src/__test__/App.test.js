import { render } from '@testing-library/react';
import App from '../App';
import axios from 'axios';

// jest.mock(axios);

test('renders without crashing', () => {
  render(<App />);
});

it('matches snapshot', () => {
  const app = render(<App />);
  expect(app.asFragment()).toMatchSnapshot();
});

it('should display title h1', () => {
  const {asFragment, getByText} = render(<App />)
  console.log(asFragment)
  expect(getByText('MorNooNightsNews')).toBeInTheDocument()
  expect(asFragment()).toMatchInlineSnapshot(`
  <h1 class="mb-4 text-white font-weight-bold">MorNooNightsNews</h1>
`)
});