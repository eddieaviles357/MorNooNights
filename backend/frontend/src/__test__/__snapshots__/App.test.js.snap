// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matches snapshot 1`] = `
<DocumentFragment>
  <div
    class="App"
  >
    <nav
      class="Nav"
    >
      <a
        class="navbar-brand"
        href="/"
      >
        MorNooNightsNews
      </a>
      <ul>
        <li>
          <a
            class=""
            href="/login"
          >
            Login
          </a>
        </li>
        <li>
          <a
            class=""
            href="/signup"
          >
            Sign Up
          </a>
        </li>
      </ul>
    </nav>
    <div
      class="Home"
    >
      <div
        class="container text-center"
      >
        <h1
          class="mb-4 text-white font-weight-bold"
        >
          MorNooNightsNews
        </h1>
        <p
          class="lead text-white"
        >
          Daily News
        </p>
        <p>
          <a
            class="Btn-tertiary btn m-3"
            href="/login"
          >
            Log in
          </a>
          <a
            class="Btn-secondary btn m-3"
            href="/signup"
          >
            Sign up
          </a>
        </p>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
